New module: remove_host

jpic opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request
 - New Module Pull Request

##### COMPONENT NAME
remove_host
##### ANSIBLE VERSION
```
ansible 2.3.0 (remove_host f7f455dd84) last updated 2016/11/23 11:21:43 (GMT +200)
  lib/ansible/modules/core: (detached HEAD b598611afb) last updated 2016/11/23 11:23:32 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD 25292b3ebd) last updated 2016/11/23 11:23:32 (GMT +200)
  config file = /home/jpic/.ansible.cfg
  configured module search path = ['/home/jpic/ansible/library', '/usr/share/ansible']
```
##### SUMMARY
Added remove_host module. This is for when a task actually destroys a host, then it can be removed from further plays.

jpic committed:
New module: remove_host

bcoca commented on lib/ansible/inventory/group.py:
```diff
@@ -114,6 +114,11 @@ def add_host(self, host):
+    def remove_host(self, host):
+        host.remove_group(self)
+        self.hosts.remove(host)
+        self.clear_hosts_cache()
```
At first I thought this might be overkill, but all hosts need to rebuild their groups vars to remove this host from them. Making note in case someone else goes down same road.

jpic replied:
Adding this as a code comment, thank you @bcoca <3

bcoca commented:
making note of needing docs (module stub) once this is merged

jpic commented:
Thanks for accepting this, but the test should fail, looks like this commit misses something to have it executed, any clue please ?

gundalow commented:
@jpic The tests need referencing from `test/integration/non_destructive.yml` (assuming they are non destructive.

jpic commented:
Thank you @gundalow ! Tests are running, but it can't find remote_host. Probably because it's missing from modules/core ? ansible/ansible-modules-core#5726

```
2016-11-25 18:10:03 ERROR! no action detected in task. This often indicates a misspelled module name, or incorrect module path.
```

Locally it works great:

```
TASK [add_host] **************************************************************************************************************************************************
task path: /home/jpic/work/novapost/ansible-boot-lxc/test.yml:8
creating host via 'add_host': hostname=testboot.lxc
changed: [localhost] => {
    "add_host": {
        "groups": [
            "testboot"
        ], 
        "host_name": "testboot.lxc", 
        "host_vars": {
            "lxc_container_config": [
                "lxc.mount.entry=/home/jpic/work/novapost/ansible-boot-lxc srv none defaults,bind,create=dir,uid=0 0 0"
            ]
        }
    }, 
    "changed": true
}

TASK [debug] *****************************************************************************************************************************************************
task path: /home/jpic/work/novapost/ansible-boot-lxc/test.yml:13
ok: [localhost] => {
    "groups": {
        "all": [
            "testboot.lxc"
        ], 
        "testboot": [
            "testboot.lxc"
        ], 
        "ungrouped": [
            "localhost"
        ]
    }
}

TASK [remove_host] ***********************************************************************************************************************************************
task path: /home/jpic/work/novapost/ansible-boot-lxc/test.yml:14
changed: [localhost] => {
    "changed": true, 
    "remove_host": {
        "host_name": "testboot.lxc"
    }
}

TASK [debug] *****************************************************************************************************************************************************
task path: /home/jpic/work/novapost/ansible-boot-lxc/test.yml:16
ok: [localhost] => {
    "groups": {
        "all": [], 
        "testboot": [], 
        "ungrouped": [
            "localhost"
        ]
    }
}
```

Isn't the best thing to do now creating another branch with a submodule reference to my fork of core modules ?

gundalow commented on test/integration/targets/remove_host/tasks/main.yml:
```diff
@@ -0,0 +1,34 @@
+# test code for the add_host action
+# (c) 2015, Matt Davis <mdavis@ansible.com>
```
Copy and paste error?

gundalow commented:
Changes to `lib/ansible/modules/*` must be in another PR, once merger we (Ansible Core Team) will update the submodules reference

Once the above is done this you can test this PR.

Please remove the change to `lib/ansible/modules/core`


The git submodule pain will be going away very shortly

gundalow commented on test/integration/targets/remove_host/tasks/main.yml:
```diff
@@ -0,0 +1,34 @@
+    that:
+    - hostvars['newdynamichost'] is not defined
+    - groups['newdynamicgroup'] is defined
+    - "not groups['newdynamicgroup']"
```
After this test it would be worth doing another `remove_host` operation, to ensure the action is idempotent.

gundalow reviewed (changes requested):
needs_revision

mattclay commented:
@jpic One of the [test failures](https://app.shippable.com/runs/584680b43ee1d30f00c9b156/1/console) is due to a yamllint test failure:

```
test/integration/targets/remove_host/tasks/main.yml
  36:25     error    syntax error: mapping values are not allowed here
```

gundalow commented:
rebase needed

jpic commented:
Done ! I'm stuck at implementing idempotence.

gundalow commented:
Regarding idempotent support, do the other modules give a clue? If not feel free to join `#ansible-devel` on Freenode IRC.

mattclay commented:
@jpic Please rewrite your last commit message to remove the copy of the playbook output. That message is preventing CI from processing your PR.

jpic commented:
Sure, the module is not detected anymore since the rebase:

```
$ make non_destructive
rm -rf ~/ansible_testing
mkdir -p ~/ansible_testing
ANSIBLE_ROLES_PATH=/home/jpic/env/src/ansible/test/integration/targets
ansible-playbook non_destructive.yml -i inventory -e
@integration_config.yml  -v
Using /home/jpic/.ansible.cfg as config file
ERROR! no action detected in task. This often indicates a misspelled module
name, or incorrect module path.

The error appears to have been in
'/home/jpic/env/src/ansible/test/integration/targets/remove_host/tasks/main.yml':
line 19, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: removing an unknown host should not do anything
  ^ here


The error appears to have been in
'/home/jpic/env/src/ansible/test/integration/targets/remove_host/tasks/main.yml':
line 19, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: removing an unknown host should not do anything
  ^ here

make: *** [Makefile:73: non_destructive] Error 4
```

mattclay commented:
@jpic You need to add a `test/integration/targets/remove_host/aliases` file to enable CI to run your new test target. At a minimum, you need to put the test into a CI group. To do that, the `aliases` file should contain:

```
posix/ci/group2
```

mattclay commented:
@jpic The `remove_host` action is present, but there's no accompanying module file to provide metadata or documentation.

mattclay commented:
CI is failing on integration tests:

```
2017-02-09 18:12:10 TASK [remove_host : removing an unknown host should not do anything] ***********
2017-02-09 18:12:10 task path: /root/ansible/test/integration/targets/remove_host/tasks/main.yml:19
2017-02-09 18:12:10 ERROR! Unexpected Exception: 'dict' object has no attribute '_result'
2017-02-09 18:12:10 the full traceback was:
2017-02-09 18:12:10 
2017-02-09 18:12:10 Traceback (most recent call last):
2017-02-09 18:12:10   File "/root/ansible/bin/ansible-playbook", line 103, in <module>
2017-02-09 18:12:10     exit_code = cli.run()
2017-02-09 18:12:10   File "/root/ansible/lib/ansible/cli/playbook.py", line 156, in run
2017-02-09 18:12:10     results = pbex.run()
2017-02-09 18:12:10   File "/root/ansible/lib/ansible/executor/playbook_executor.py", line 153, in run
2017-02-09 18:12:10     result = self._tqm.run(play=play)
2017-02-09 18:12:10   File "/root/ansible/lib/ansible/executor/task_queue_manager.py", line 284, in run
2017-02-09 18:12:10     play_return = strategy.run(iterator, play_context)
2017-02-09 18:12:10   File "/root/ansible/lib/ansible/plugins/strategy/linear.py", line 278, in run
2017-02-09 18:12:10     results += self._wait_on_pending_results(iterator)
2017-02-09 18:12:10   File "/root/ansible/lib/ansible/plugins/strategy/__init__.py", line 581, in _wait_on_pending_results
2017-02-09 18:12:10     results = self._process_pending_results(iterator)
2017-02-09 18:12:10   File "/root/ansible/lib/ansible/plugins/strategy/__init__.py", line 487, in _process_pending_results
2017-02-09 18:12:10     self._remove_host(old_host_info, iterator, result_item)
2017-02-09 18:12:10   File "/root/ansible/lib/ansible/plugins/strategy/__init__.py", line 602, in _remove_host
2017-02-09 18:12:10     result_item._result['changed'] = False
2017-02-09 18:12:10 AttributeError: 'dict' object has no attribute '_result'
```

jpic commented:
Thanks, I've finished the rebase. However, I'm a bit stuck at passing the test:

```
- name: remove unknown host
  remove_host:
    name: removedhost
  register: remove_host_idempotence

- debug: var=remove_host_idempotence
```

While I get the changed=false as expected from the remove_host output, the debug module call shows changed=true !

```

    TASK [remove_host : remove unknown host] ******************************************************************************************************************************************************
    task path: /home/jpic/env/src/ansible/test/integration/targets/remove_host/tasks/main.yml:46
    ok: [testhost] => {
        "changed": false,
        "remove_host": {
            "host_name": "removedhost"
        }
    }

    TASK [remove_host : debug] ********************************************************************************************************************************************************************
    task path: /home/jpic/env/src/ansible/test/integration/targets/remove_host/tasks/main.yml:51
    ok: [testhost] => {
        "changed": false,
        "remove_host_idempotence": {
            "changed": true, # <---- WOOOOOOOOOOOOOT ?
            "remove_host": {
                "host_name": "removedhost"
            }
        }
    }
```

mattclay commented:
@jpic Where you seeing that error? The only CI error I see is the following:

```
2017-02-16 16:12:39 ============================================================================
2017-02-16 16:12:39 lib/ansible/modules/inventory/remove_host.py
2017-02-16 16:12:39 ============================================================================
2017-02-16 16:12:39 ERROR: ANSIBLE_METADATA.version: not a valid value for dictionary value @ data['version']. Got '2.4.0'
```

gundalow commented:
Rebase needed

gundalow commented:
So apparently we (Core) want this, though unsure about functionality so I'm leaving this open and sticking P3 label on.

jpic commented:
Sorry for the delay Core, this completely went under my radar.
I just rebased, feel free to ask me for changes or close.

jpic commented:
It looks like CI is stale is that possible ? I have fixed the last ansibot comment I think

tremble commented on test/integration/targets/remove_host/tasks/main.yml:
```diff
@@ -0,0 +1,40 @@
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+- name: removing an unknown host should not fail
+  remove_host: name=remove_host_unknown_host
```
Should we be asserting `is not changed` here?

jpic replied:
Well, that's what did the initial PR, but that's not possible anymore due to internal changes, the same way it's not possible for add_host:

```
$ cat test.yml 
#!/usr/bin/env ansible-playbook
- hosts: '*'
  tasks:
  - add_host: name=bar
    register: add_host0
  - debug: var=add_host0
  - add_host: name=bar
    register: add_host1
  - debug: var=add_host1


$ ansible-playbook  -i localhost, -c local test.yml 

PLAY [*] ******************************************************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************
ok: [localhost]

TASK [add_host] ***********************************************************************************************************************************
changed: [localhost]

TASK [debug] **************************************************************************************************************************************
ok: [localhost] => {
    "add_host0": {
        "add_host": {
            "groups": [],
            "host_name": "bar",
            "host_vars": {}
        },
        "changed": true,
        "failed": false
    }
}

TASK [add_host] ***********************************************************************************************************************************
changed: [localhost]

TASK [debug] **************************************************************************************************************************************
ok: [localhost] => {
    "add_host1": {
        "add_host": {
            "groups": [],
            "host_name": "bar",
            "host_vars": {}
        },
        "changed": true,
        "failed": false
    }
}

PLAY RECAP ****************************************************************************************************************************************
localhost                  : ok=5    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

I did debug that to figure why:

- there seems to be now way for the StrategyBase to communicate with the ActionModule,
- even when `result['changed']` is not explicitly set to True in the ActionModule, it will be set to True by default,
- changing `result_item['changed']` in the StrategyBase (which is what this PR did initially) has no effect on the returned result by the ActionModule anymore

tremble commented on test/integration/targets/remove_host/tasks/main.yml:
```diff
@@ -0,0 +1,40 @@
+
+- name: remove host from the runtime inventory
+  remove_host:
+    name: removedhost
```
And `similarly is changed`

jpic replied:
Same issue as above unfortunately

jpic commented:
@gundalow is there any change request that is not implemented ?
